src/main/java/org/opensearch/ad/ml/CheckpointDao.java

```diff
@@ -298,16 +332,39 @@ public String toCheckpoint(EntityModel model) {
         });
     }
 
-    /**
-     * Deletes the model checkpoint for the id.
-     *
-     * @deprecated use deleteModelCheckpoint with listener instead
-     *
-     * @param modelId ID of the model checkpoint
-     */
-    @Deprecated
-    public void deleteModelCheckpoint(String modelId) {
-        clientUtil.<DeleteRequest, DeleteResponse>timedRequest(new DeleteRequest(indexName, modelId), logger, client::delete);
+    private String rcfModelToCheckpoint(RandomCutForest model) {
+        LinkedBuffer borrowedBuffer = null;
+        try {
+            borrowedBuffer = deserializeRCFBufferPool.borrowObject();
+            try {
+                return rcfModelToCheckpoint(model, borrowedBuffer);
+            } catch (Exception e) {
+                if (borrowedBuffer != null) {
+                    deserializeRCFBufferPool.invalidateObject(borrowedBuffer);
+                    borrowedBuffer = null;
+                }
+            } finally {
+                if (borrowedBuffer != null) {
+                    deserializeRCFBufferPool.returnObject(borrowedBuffer);
+                }
+            }
+        } catch (Exception e) {
+            logger.error("Failed to borrow an buffer from object pool", e);
```

Comment:
Is this an error if we continue to execute? Are there any valid cases where borrowedBuffer should be expected to be null? In line 359 we allocate a new LinkedBuffer.

Reply:
What I wrote is if we cannot borrow any buffer from object pool then go to allocate a new LinkedBuffer in line 359.

We return null when we successfully borrow an object but fail serialization.  I moved return null to the exception block to be more explicit and added more comments.

```
private String rcfModelToCheckpoint(RandomCutForest model) {
        LinkedBuffer borrowedBuffer = null;
        try {
            borrowedBuffer = deserializeRCFBufferPool.borrowObject();
            try {
                return rcfModelToCheckpoint(model, borrowedBuffer);
            } catch (Exception e) {
                if (borrowedBuffer != null) {
                    deserializeRCFBufferPool.invalidateObject(borrowedBuffer);
                    borrowedBuffer = null;
                }
                // return null when we successfully borrow an object but fail serialization
                return null;
            } finally {
                if (borrowedBuffer != null) {
                    deserializeRCFBufferPool.returnObject(borrowedBuffer);
                }
            }
        } catch (Exception e) {
            logger.error("Failed to borrow an buffer from object pool", e);
            return rcfModelToCheckpoint(model, null);
        }
    }
```